### PR TITLE
Adding SimpleITK recipe

### PR DESCRIPTION
**Pull request by blowekamp**
Moving recipe from conda-recipe repository.

This package was manually packaged across platforms and placed on
anaconda cloud in the "simpleitk" channel.

### CODE REVIEW

**Comment by patricksnape** [comment]
Sorry, there was a problem with the CI server and so `hachoir-core` and `hyperspy` were not removed from the repo. Could you please rebase this PR on top of master (I just fixed it)?

**Comment by blowekamp** [comment]
There are failures in compilation due to cmake missing openssl ( conda-forge/cmake-feedstock#7 )

File: recipes/simpleitk/bld.bat [hunk 1]
```diff
@@ -0,0 +1,49 @@
+mkdir C:\b
+cd C:\b
+
+set "CMAKE_GENERATOR=Visual Studio 9 2008"
+if "%PY_VER%" == "3.3" set "CMAKE_GENERATOR=Visual Studio 10"
+if "%PY_VER%" == "3.4" set "CMAKE_GENERATOR=Visual Studio 10"
+if "%PY_VER%" == "3.5" set "CMAKE_GENERATOR=Visual Studio 14"
+
+if "%ARCH%" == "64" set "CMAKE_GENERATOR=%CMAKE_GENERATOR% Win64"
```
Comment (jakirkham) [inline recipes/simpleitk/bld.bat]: So, I think `conda-build` sets this for you. Is that right, @patricksnape? I ask you as I thought you added it. 😄

File: recipes/simpleitk/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    number: 0
+
+requirements:
+  build:
+    - python
+    - cmake    >=2.8.11
+    - libtiff  4.0.*  # [not win]
+    - libpng   1.6.*  # [not win]
+    - jpeg     8*  # [not win]
+    - zlib     1.*  # [not win]
```
Comment (jakirkham) [inline recipes/simpleitk/meta.yaml]: Please update to match these [pinnings]( https://github.com/conda-forge/conda-forge.github.io/blob/df07f63f6c7729d1bc96177f7de85b991483e819/scripts/pin_the_slow_way.py#L32-L48 ).

File: recipes/simpleitk/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    - setuptools
+  run:
+    - python
+    - libtiff  4.0.*  # [not win]
+    - libpng   1.6.*  # [not win]
+    - jpeg     8*  # [not win]
+    - zlib     1.*  # [not win]
```
Comment (jakirkham) [inline recipes/simpleitk/meta.yaml]: Same as above.

File: recipes/simpleitk/build.sh [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    -D "PYTHON_EXECUTABLE:FILEPATH=${PYTHON}" \
+    -D "PYTHON_INCLUDE_DIR:PATH=$PREFIX/include/python${MY_PY_VER}" \
+    -D "PYTHON_LIBRARY:FILEPATH=$PREFIX/lib/libpython${MY_PY_VER}.${SO_EXT}" \
+    "${SRC_DIR}/SuperBuild"
+
+make -j ${CORES}
+cd ${BUILD_DIR}/SimpleITK-build/Wrapping
+${PYTHON} PythonPackage/setup.py install
+
+
```
Comment (jakirkham) [inline recipes/simpleitk/build.sh]: nit: please strip extra newline. one terminal newline is sufficient.

**Comment by jakirkham** [comment]
Have you tried using [`ninja`]( http://ninja-build.org/ )? That may speed up the build times. It's easy to get working with `cmake`.

**Comment by blowekamp** [comment]
Thank you for those comments. I have made the changes and am waiting until for cmake to be updated with ssl before I push.

Ninja does improve the performance some, but it is not going to be enough to complete all the builds in the allocated time. Also there have been some recent changes to SimpleITK's CMake to help with the ninja building, these are in the current 0.10rc1.

**Comment by jakirkham** [comment]
So, the `cmake` build has been deployed. Please let me know if you run into other issues.

File: recipes/simpleitk/bld.bat (hunk 1 above)
Comment (patricksnape) [inline recipes/simpleitk/bld.bat]: Yes it does: [docs](http://conda.pydata.org/docs/building/environment-vars.html)

**Comment by jakirkham** [comment]
There was a dependency issue that needed to be resolved. We are currently deploying a new `cmake` to fix that.

File: recipes/simpleitk/build.sh [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+cd ${BUILD_DIR}
+
+MY_PY_VER=${PY_VER}
+if [ $PY3K -eq "1" ]; then
+    MY_PY_VER="${MY_PY_VER}m"
+fi
+
+
+cmake \
+    -G "%CMAKE_GENERATOR%" \
```
Comment (blowekamp) [inline recipes/simpleitk/build.sh]: This line will not work!

Comment (jakirkham) [inline recipes/simpleitk/build.sh]: I think it can't be quoted.

**Comment by blowekamp** [comment]
The CMake SSL issue seems resolved.

The Travis build completed Python 2.7, and timed out with python 3.4.
CircleCi and Appveyor timeout with Python 2.7

@msarahan

**Comment by jakirkham** [comment]
It appears the compiler was crashed a few times in CircleCI too.

**Comment by blowekamp** [comment]
@jakirkham I thought the compiler termination was just how the CI service terminated the process for a time out?

I just switched to using Ninja, and Ninja is not producing progress, so the builds timed out due to lack of output.

**Comment by jakirkham** [comment]
Try adding `unset MAKEFLAGS` to `build.sh`.

**Comment by blowekamp** [comment]
What will unsetting MAKEFLAGS do?

**Comment by ocefpaf** [comment]
> What will unsetting MAKEFLAGS do?

Fixes a bug in `conda-forge`'s config where it forces a flag everywhere even though it is useful only in a few cases.

**Comment by jakirkham** [comment]
`MAKEFLAGS` currently is used to set the number of cores to build with. Unfortunately it happens without the users knowledge, which can be a bit surprising.

Given the compiler is crashing on CircleCI, I can only assume that we do not have the degree of parallelism building this as the number of cores provided by CircleCI would imply.

**Comment by blowekamp** [comment]
I think that the builds may finish one package within the time constraint. What would be the best way to get the CI to only try to build python 3.5?

**Comment by kmuehlbauer** [comment]
ping @blowekamp @conda-forge/core

Is there any progress planned in adding this recipe?

File: .CI/build_all [hunk 6]
```diff
@@ -24,5 +24,5 @@ if [ -z "$GH_TOKEN" ]; then
     rm -rf ./recipes/example
 
     # We just want to build all of the recipes.
-    conda-build-all ./recipes --matrix-condition "numpy >=1.10" "python >=2.7,<3|>=3.4"
+    conda-build-all ./recipes --matrix-condition "numpy >=1.10" "python 3.5.*"
```
Comment (ocefpaf) [inline .CI/build_all]: You should not change the CI configuration. If you want to build only for `Python 3.5` you need to add a skip statement in the recipe.

File: appveyor.yml [hunk 7]
```diff
@@ -11,26 +11,6 @@ environment:
   matrix:
     # Note: Because we have to separate the py2 and py3 components due to compiler version, we have a race condition for non-python packages.
     # Not sure how to resolve this, but maybe we should be tracking the VS version in the build string anyway?
-    - TARGET_ARCH: "x86"
```
Comment (ocefpaf) [inline appveyor.yml]: Again, do not change the CI configuration.

File: recipes/simpleitk/bld.bat [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    -D "PYTHON_INCLUDE_DIR:PATH=%PREFIX%/include" ^
+    -D "PYTHON_LIBRARY:FILEPATH=%PREFIX%/libs/python%MY_PY_VER%.lib" ^
+    "%SRC_DIR%/SuperBuild"
+
+if errorlevel 1 exit 1
+
+REM Build step
+cmake --build  . --config Release
+if errorlevel 1 exit 1
+
```
Comment (ocefpaf) [inline recipes/simpleitk/bld.bat]: Is there a `ctest` that we can run here?

File: recipes/simpleitk/build.sh [hunk 9]
```diff
@@ -0,0 +1,53 @@
+#!/bin/bash
+
+CORES=$(grep -c ^processor /proc/cpuinfo 2>/dev/null || sysctl -n hw.ncpu || 1)
+
+if [ `uname` == Darwin ]; then
+    SO_EXT='dylib'
+    CMAKE_ARGS="-D CMAKE_OSX_DEPLOYMENT_TARGET:STRING=${MACOSX_DEPLOYMENT_TARGET}"
+else
+    SO_EXT='so'
```
Comment (ocefpaf) [inline recipes/simpleitk/build.sh]: Please use the `SHLIB_EXT` instead. See the `openjpeg` feedstock for an example.

File: recipes/simpleitk/build.sh [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    -D ITK_USE_SYSTEM_ZLIB:BOOL=ON \
+    -D USE_SYSTEM_SWIG:BOOL=ON \
+    -D "CMAKE_SYSTEM_PREFIX_PATH:FILEPATH=${PREFIX}" \
+    -D "PYTHON_EXECUTABLE:FILEPATH=${PYTHON}" \
+    -D "PYTHON_INCLUDE_DIR:PATH=$PREFIX/include/python${MY_PY_VER}" \
+    -D "PYTHON_LIBRARY:FILEPATH=$PREFIX/lib/libpython${MY_PY_VER}.${SO_EXT}" \
+    "${SRC_DIR}/SuperBuild"
+
+make -j ${CORES}
+cd ${BUILD_DIR}/SimpleITK-build/Wrapping/Python
```
Comment (ocefpaf) [inline recipes/simpleitk/build.sh]: Is there a `make check` we can run?

File: recipes/simpleitk/meta.yaml [hunk 11]
```diff
@@ -0,0 +1,43 @@
+{% set version = "0.10.0" %}
+
+package:
+  name: simpleitk
+  version: "{{version}}"
```
Comment (ocefpaf) [inline recipes/simpleitk/meta.yaml]: No need for the quotes.

File: recipes/simpleitk/meta.yaml [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+
+source:
+  git_url: http://itk.org/SimpleITK.git
```
Comment (ocefpaf) [inline recipes/simpleitk/meta.yaml]: Please change this to a tarball. For example:

```yaml
source:
  fn: seawater-{{ version }}.tar.gz
  url: https://github.com/pyoceans/python-seawater/archive/v{{ version }}.tar.gz
  sha256: 8fecf37d69fbe5f2bf50855aab38f9438b44ee4195e925edfa9e544f75cfecc8
```

File: recipes/simpleitk/meta.yaml [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+
+requirements:
+  build:
+    - swig
+    - python
+    - cmake    >=2.8.11
+    - libtiff  4.0.*
+    - libpng   >=1.6.21,<1.7
+    - jpeg     8*
+    - zlib     1.*
```
Comment (ocefpaf) [inline recipes/simpleitk/meta.yaml]: Update all the pinnings based on https://github.com/conda-forge/staged-recipes/wiki/Pinned-dependencies

File: recipes/simpleitk/meta.yaml [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+    - setuptools
+  run:
+    - python
+    - libtiff  4.0.*
+    - libpng   >=1.6.21,<1.7
+    - jpeg     8*
+    - zlib     1.*
```
Comment (ocefpaf) [inline recipes/simpleitk/meta.yaml]: Same as https://github.com/conda-forge/staged-recipes/pull/1045/files#r98656952

File: scripts/run_docker_build.sh [hunk 15]
```diff
@@ -48,6 +48,6 @@ find conda-recipes -mindepth 2 -maxdepth 2 -type f -name "yum_requirements.txt"
     | xargs -n1 cat | grep -v -e "^#" -e "^$" | \
     xargs -r yum install -y
 
-conda-build-all /conda-recipes --matrix-conditions "numpy >=1.10" "python >=2.7,<3|>=3.4"
+conda-build-all /conda-recipes --matrix-conditions "numpy >=1.10" "python 3.5.*"
```
Comment (ocefpaf) [inline scripts/run_docker_build.sh]: Ditto.

File: recipes/simpleitk/bld.bat (hunk 8 above)
Comment (blowekamp) [inline recipes/simpleitk/bld.bat]: Do you mean ctest to build?  It my experience that the ctest program is used to drive a testing cycle of Start, Update, Configure, Build, Test, Submit. Here what are just trying to do the equivalent of "make", with the specified generator. So I think this is appropriate. If you have a link to information on your expected behavior please provide the further explanation.

File: recipes/simpleitk/build.sh (hunk 9 above)
Comment (blowekamp) [inline recipes/simpleitk/build.sh]: Great!

File: recipes/simpleitk/build.sh (hunk 10 above)
Comment (blowekamp) [inline recipes/simpleitk/build.sh]: There is a test suit. That can be run with ctest.. but testing is disabled... and we are timing out.

File: recipes/simpleitk/meta.yaml (hunk 12 above)
Comment (blowekamp) [inline recipes/simpleitk/meta.yaml]: OK

**Comment by blowekamp** [comment]
Last this was tried, we were timing out. There was an attempt made to see if we could get just one Python version compiled in the time allocated. The time out appears to be a blocker, and we are hoping that the policies and resources would evolve.

An alternative approach we were considering is to break this project in two, a C++ "core" SimpleITK library, and then the Python wrapping.  The "core" library would be static C++ libraries and C++ headers. This would be very beneficial to linux where the same C++ "core" library currently has to get compiled for each Python version. Then a separate project  would the SWIG wrapping for the different Python versions/languages. The SimpleITK CMake build infrastructure does support this, but we have not invested the time with the conda-build system to evaluate this.

Currently, @zivy has worked on a [recipe](https://github.com/SimpleITK/SimpleITKCondaRecipe) that we should integrate these updates into.

File: recipes/simpleitk/bld.bat [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+    -D ITK_USE_SYSTEM_JPEG:BOOL=ON ^
+    -D ITK_USE_SYSTEM_PNG:BOOL=ON ^
+    -D ITK_USE_SYSTEM_TIFF:BOOL=ON ^
+    -D ITK_USE_SYSTEM_ZLIB:BOOL=ON ^
+    -D "CMAKE_SYSTEM_PREFIX_PATH:PATH=%PREFIX%/Library" ^
+    -D "PYTHON_EXECUTABLE:FILEPATH=%PYTHON%" ^
+    -D "PYTHON_INCLUDE_DIR:PATH=%PREFIX%/include" ^
+    -D "PYTHON_LIBRARY:FILEPATH=%PREFIX%/libs/python%MY_PY_VER%.lib" ^
+    "%SRC_DIR%/SuperBuild"
+
```
Comment (ocefpaf) [inline recipes/simpleitk/bld.bat]: I believe you need to remove this blank line or the line below won't work.

File: recipes/simpleitk/bld.bat (hunk 8 above)
Comment (ocefpaf) [inline recipes/simpleitk/bld.bat]: See See this example: https://github.com/conda-forge/libglu-feedstock/blob/master/recipe/bld.bat
All I mean is that if the tests exist, please tun them.

**Comment by kmuehlbauer** [comment]
> An alternative approach we were considering is to break this project in two, a C++ "core" SimpleITK library, and then the Python wrapping. The "core" library would be static C++ libraries and C++ headers. This would be very beneficial to linux where the same C++ "core" library currently has to get compiled for each Python version. Then a separate project would the SWIG wrapping for the different Python versions/languages. The SimpleITK CMake build infrastructure does support this, but we have not invested the time with the conda-build system to evaluate this.

Thanks for the heads-up @blowekamp. If I understand correctly you are suggesting to have SimpleITK and python-SimpleITK packages? Is a separate SWIG wrapping project really necessary? There are other examples of C/C++-libraries (which bring their own wrappers) having feedstocks for the core-lib and several other feedstocks for the bindings (eg. python, fortran). @ocefpaf could this be done the same way here (like netcdf, eccodes...)?

File: recipes/simpleitk/build.sh (hunk 10 above)
Comment (ocefpaf) [inline recipes/simpleitk/build.sh]: Can you add a comment in the recipe for that? The tests may pass in the feedstock or at least for Circle-CI where we get more time.

**Comment by ocefpaf** [comment]
> @ocefpaf could this be done the same way here (like netcdf, eccodes...)?

Sure.

**Comment by blowekamp** [comment]
The SimpleITK development recipe is proposed here: #2846 2846

**Comment by blowekamp** [comment]
Creating a new PR.

**Pull request closed without merging**